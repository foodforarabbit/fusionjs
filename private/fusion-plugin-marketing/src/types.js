// @flow

import type {Context} from 'fusion-core';

export type PluginConfig = {
  cookieAge: number,
  cookieIdKey: string,
  cookieDomain: string | null,
  serverDomain: string,
  disableHeatpipe: boolean,
  debugLogging: boolean,
};

export type PluginServiceType = {
  from: (
    ctx: Context
  ) => {
    ctx: Context,
    config: PluginConfig,
    getAnalyticsSessionId: (ctx?: Context) => string,
    getCookieId: () => string,
    setCookieId: () => void,
    shouldSkipTracking: () => boolean,
    publish: () => Promise<void>,
    handleClientTrackRequest: () => TrackResponse,
    setTrackingInfo: (ctx?: Context) => void,
    track: (ctx?: Context) => void,
  },
};

export type TrackResponse = {
  cookie_id: string,
  session_id: string,
};

export type BrowserPluginServiceType = {
  from: (
    ctx: Context
  ) => {
    getTrackResponse: () => ?TrackResponse,
    track: () => void,
  },
};

export type PluginCanActivate = {
  canActivate: boolean,
};
