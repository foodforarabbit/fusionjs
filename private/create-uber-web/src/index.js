/* @flow */

import sade from 'sade';
import {scaffold} from './commands/scaffold.js';
import {upgrade} from './commands/upgrade.js';
import {provision} from './commands/provision.js';

const whitelist = 'fusion|@uber|styletron|eslint|flow-bin';

const cli = sade('uber-web');

cli.version(require('../package.json').version);

cli
  .command('scaffold')
  .describe('Scaffold a new project structure locally')
  .option('--type', 'website, fusion-plugin or library', '')
  .option('--name', 'Project name', '')
  .option('--description', 'Project description', '')
  .option('--team', 'Team name', '')
  .option('--audience', 'external or internal')
  .option('--local-path, -l', 'Use a local folder as the template')
  .option('--skip-install', 'Avoid running `yarn install`')
  .option('--hoist-deps', [
    'Generates a package.json without dependencies fields',
    'Use this flag if scaffolding into a versionless monorepo with a global yarn.lock file',
  ])
  .action(args => {
    const {
      type,
      name,
      description,
      team,
      audience,
      l: localPathShorthand,
      'local-path': localPath,
      'skip-install': skipInstall = false,
      'hoist-deps': hoistDeps = false,
    } = args;
    scaffold({
      type,
      name,
      description,
      team,
      external: {external: true, internal: false}[audience],
      localPath: localPath || localPathShorthand,
      skipInstall,
      hoistDeps,
    });
  });

cli
  .command('upgrade')
  .describe('Upgrade dependencies')
  .option('--dir', 'Project folder', '.')
  .option('--match', 'Only upgrade deps whose name match this regex', whitelist)
  .option('--codemod', 'Also run Fusion.js codemods?', 'true')
  .option('--force', 'Skip tests', 'true')
  .option('--edge', 'Upgrade to prerelease if available', 'false')
  .action(args => {
    const {dir, match, codemod, force, edge} = args;
    upgrade({
      dir,
      match,
      codemod: codemod === 'true',
      force: force !== 'false',
      edge: edge !== 'false',
    });
  });

cli
  .command('provision')
  .describe('Publish a web application')
  .action(provision);

if (process.argv.length === 2) process.argv.push('--help');
cli.parse(process.argv);
