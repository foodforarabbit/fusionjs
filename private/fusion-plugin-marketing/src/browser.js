// @flow
/* eslint-env browser */

import {createPlugin} from 'fusion-core';
import type {Context, FusionPlugin} from 'fusion-core';
import type {PluginServiceType} from './types.js';

const plugin =
  __BROWSER__ &&
  createPlugin({
    provides() {
      return {
        from: (ctx?: Context) => {
          throw new Error(
            '[fusion-plugin-marketing] No service avaiable in browser.'
          );
        },
      };
    },
  });

export default ((plugin: any): FusionPlugin<empty, PluginServiceType>);
