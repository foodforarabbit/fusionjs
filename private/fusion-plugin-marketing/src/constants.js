// @flow
export const TRACK_URL = '/_track';

export const QUERY_KEYS = [
  'uclick_id',
  'utm_source',
  'utm_term',
  'utm_content',
  'utm_medium',
  'utm_campaign',
  'ad_id',
  'adgroup_id',
  'campaign_id',
  'kw_id',

  'gclid',
  'gclsrc',
  'click_id',
  '_ga',
  'lang',
  'exp_var',
  'exp',
  'lic',
  'invite_code',
  'locality',
  'city_id',
  'ext_id',
  'ran',
  'lint_id',
  'lphy_id',
  'pos',
  'dev',
  'net',
  'match',
  'placement',
  'target',

  // Deprecated
  'ag_kwid',
  'kwid',
  'kw',
  'cid',
  'adg_id',
];

export const TRACK_TOPIC = {
  version: 11,
  topic: 'hp-clay_marketing_tracking-track',
};

export const OPTIMIZELY_COOKIE_KEY = 'optimizely-pid_cid_exp_var';
