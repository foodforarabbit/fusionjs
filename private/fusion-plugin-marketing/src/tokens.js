// @flow
import {createToken} from 'fusion-core';
import type {
  PluginConfig,
  PluginServiceType,
  PluginCanActivate,
  BrowserPluginServiceType,
} from './types';

export const UberMarketingToken = createToken<PluginServiceType>(
  'UberMarketingToken'
);
export const UberMarketingBrowserToken = createToken<BrowserPluginServiceType>(
  'UberMarketingBrowserToken'
);
export const UberMarketingConfigToken = createToken<PluginConfig>(
  'UberMarketingConfigToken'
);

export const UberMarketingCanActivateToken = createToken<PluginCanActivate>(
  'UberMarketingCanActivateToken'
);
