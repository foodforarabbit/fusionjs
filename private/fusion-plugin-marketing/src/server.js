// @flow

import {createPlugin, memoize} from 'fusion-core';
import stringify from 'fast-safe-stringify';
import type {FusionPlugin, Context} from 'fusion-core';
import type {PluginServiceType, PluginConfig} from './types.js';

import {LoggerToken} from 'fusion-tokens';
import {HeatpipeToken} from '@uber/fusion-plugin-heatpipe';

import {TRACK_TOPIC} from './constants';

import {getCookieId, setCookieId} from './utils/cookies';
import {createTrackingPayload, shouldSkipTracking} from './utils/tracking';

import {UberMarketingConfigToken} from './tokens';

const plugin =
  __NODE__ &&
  createPlugin({
    deps: {
      userConfig: UberMarketingConfigToken,
      heatpipe: HeatpipeToken,
      logger: LoggerToken,
    },
    provides({userConfig = {}}) {
      class MarketingPlugin {
        ctx: Context;
        config: PluginConfig;
        visitorUuid: string;

        constructor(ctx) {
          this.ctx = ctx;
          this.config = {
            cookieAge: 31536000,
            cookieIdKey: 'marketing_vistor_id',
            cookieDomain: __DEV__ ? 'localhost' : '.uber.com',
            serverDomain: '', // TODO: consolidate this to a single meta field
            disableHeatpipe: false,
            ...userConfig,
          };

          if (!this.config.serverDomain) {
            throw new Error(
              '[fusion-plugin-marketing] MISSING config.serverDomain = YOUR_WEB_APP_DOMAIN'
            );
          }

          this.visitorUuid = getCookieId(ctx, this.config.cookieIdKey);
        }

        getCookieId() {
          return this.visitorUuid;
        }
      }
      return {
        from: memoize(ctx => new MarketingPlugin(ctx)),
      };
    },
    middleware({userConfig = {}, heatpipe, logger}, MarketingPlugin) {
      return async (ctx: Context, next) => {
        const marketing = MarketingPlugin.from(ctx);
        const {config} = marketing;

        setCookieId(ctx, marketing.getCookieId(), config);
        await next();

        ctx.timing.end.then(() => {
          // Publish tracking during request upstream for the status code
          if (!config.disableHeatpipe) {
            try {
              const trackingInfo = createTrackingPayload(
                ctx,
                marketing.getCookieId(),
                config
              );
              if (!shouldSkipTracking(ctx, trackingInfo)) {
                logger.info('[fusion-plugin-marketing] Tracking', {
                  data: stringify(trackingInfo),
                });
                heatpipe.publish(TRACK_TOPIC, trackingInfo);
              }
            } catch (e) {
              logger.error('[fusion-plugin-marketing] Unexpected Error', {
                error: e,
              });
            }
          }
        });
      };
    },
  });

export default ((plugin: any): FusionPlugin<empty, PluginServiceType>);
