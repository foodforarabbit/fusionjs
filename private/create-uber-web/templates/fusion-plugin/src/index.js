// @flow
/* eslint-env browser */
import {createToken} from 'fusion-core';
import server from './server.js';
import browser from './browser.js';

export default (__NODE__ ? server : browser);
