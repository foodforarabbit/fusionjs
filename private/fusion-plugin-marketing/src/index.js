// @flow
/* eslint-env browser */
import server from './serverPlugin/server.js';
import serverWarn from './serverPlugin/browser.js';

import browserWarn from './MarketingBrowserPlugin/server.js';
import browser from './MarketingBrowserPlugin/browser.js';

export const MarketingBrowserPlugin = __NODE__ ? browserWarn : browser;
export default __NODE__ ? server : serverWarn;
export {
  UberMarketingToken,
  UberMarketingBrowserToken,
  UberMarketingConfigToken,
  UberMarketingCanActivateToken,
} from './tokens';
