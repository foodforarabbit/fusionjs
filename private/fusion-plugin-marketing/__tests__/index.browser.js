// @flow

/* global global */

import App, {createPlugin} from 'fusion-core';
import {getSimulator} from 'fusion-test-utils';
import {FetchToken} from 'fusion-tokens';

import {UberMarketingBrowserToken} from '../src/tokens';
import Plugin from '../src/MarketingBrowserPlugin/browser';

let windowSpy;

beforeEach(() => {
  windowSpy = jest.spyOn(global, 'window', 'get');
});

afterEach(() => {
  windowSpy.mockRestore();
});

test('test', () => {
  expect(typeof window).toEqual('object');
});

const fetchMock = jest.fn().mockImplementation(() => {
  return Promise.resolve({
    json: () => {
      return {
        session_id: 'mock_session_id',
        cookie_id: 'mock_cookie_id',
      };
    },
  });
});

function createMockApp({testPlugin}: {testPlugin?: Object}) {
  const app = new App({}, () => 'ok');
  // $FlowFixMe
  app.register(testPlugin);
  // $FlowFixMe
  app.register(FetchToken, fetchMock);
  app.register(UberMarketingBrowserToken, Plugin);

  const simulator = getSimulator(app);
  return simulator;
}

test('should POST to /_track with query string and referrer', async done => {
  const mockPath =
    '/app?utm_source=foo&utm_campaign=bar&utm_campaign=not-picked';
  const testPlugin = createPlugin({
    deps: {
      Marketing: UberMarketingBrowserToken,
    },
    middleware({Marketing}) {
      return async (ctx, next) => {
        const marketing = Marketing.from(ctx);
        const res = await marketing.getTrackResponse();
        setTimeout(() => {
          expect(fetchMock).toHaveBeenCalled();
          expect(fetchMock.mock.calls[0]).toMatchSnapshot(
            'fetch request params'
          );
          expect(res).toMatchSnapshot('fetch response');
          done();
        }, 100);
      };
    },
  });

  const location = {
    href: `https://www.uber.com${mockPath}`,
  };

  const document = {
    location,
    referrer: 'https://www.uber.com/referrer',
  };

  const window = {
    location,
    document,
  };
  windowSpy.mockImplementation(() => window);
  const simulator = createMockApp({testPlugin});
  await simulator.render(mockPath);
});
