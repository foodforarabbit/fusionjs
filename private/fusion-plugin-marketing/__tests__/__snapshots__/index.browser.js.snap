// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should POST to /_track with query string and referrer: fetch request params 1`] = `
Array [
  "/_track?utm_source=foo&utm_campaign=bar&utm_campaign=not-picked",
  Object {
    "body": "{\\"referrer\\":\\"https://www.uber.com/referrer\\"}",
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`should POST to /_track with query string and referrer: fetch response 1`] = `
Object {
  "cookie_id": "mock_cookie_id",
  "session_id": "mock_session_id",
}
`;
